### PR TITLE
Feature/liquidity minting issue time adjustment

### PR DESCRIPTION
**Pull request by stanieltron**
-Pallet issuance is now responsible for distribution of amounts to pools dedicated for liquidity mining rewards
-Added sudo fn promote pool, which makes pool eligible for liquidity mining rewards
-Added functionality around calculating work when pool is promoted after it has some liquidity
-Unit tests

API change. RCP call calculate_rewards_amount now returns (not_yet_claimed, to_be_claimed) reward amounts, so users total claimable reward is not_yet_claimed + to_be_claimed

### CODE REVIEW

File: pallets/xyk/src/lib.rs [hunk 1]
```diff
@@ -569,39 +592,100 @@ pub mod pallet {
...
 
 impl<T: Config> Pallet<T> {
-	pub fn calculate_available_rewards_for_pool(
+	pub fn get_liquidity_mining_split() -> sp_runtime::Perbill {
+		<T as Config>::LiquidityMiningSplit::get_liquidity_mining_split()
+	}
+
+	pub fn get_linear_issuance_blocks() -> u32 {
+		<T as Config>::LinearIssuanceBlocks::get_linear_issuance_blocks()
+	}
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:614]: seems to be unused, please remove

File: pallets/xyk/src/lib.rs [hunk 2]
```diff
@@ -269,7 +269,10 @@ const TREASURY_PERCENTAGE: u128 = 5;
 const BUYANDBURN_PERCENTAGE: u128 = 5;
 const FEE_PERCENTAGE: u128 = 30;
 const POOL_FEE_PERCENTAGE: u128 = FEE_PERCENTAGE - TREASURY_PERCENTAGE - BUYANDBURN_PERCENTAGE;
-const TIMEBLOCKRATIO: u32 = 1000;
+// Smallest granularity in which values for liquidity minting rewards change
+const TIMEBLOCKRATIO: u32 = 10000;
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:273]: maybe that value should be passed as one of xyk config parameter

sth lik https://github.com/mangata-finance/mangata-node/blob/b90f22213d6cc120078a5007dc6423549b8a6104/pallets/xyk/src/lib.rs#L303

File: pallets/xyk/src/lib.rs [hunk 3]
```diff
@@ -569,39 +592,100 @@ pub mod pallet {
...
+				Balance::try_from(0).unwrap(),
+				liquidity_assets_amount.into(),
+			)?;
 
-		Ok(available_rewards_for_pool)
+			LiquidityMiningUser::<T>::try_get((&user, &liquidity_asset_id))
+				.unwrap_or_else(|_| (0, U256::from(0), U256::from(0)));
+
+			Ok((current_time, user_work_total, user_missing_at_checkpoint))
+		}
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:656]:
```
		match LiquidityMiningUser::<T>::try_get((&user, &liquidity_asset_id)) {
			Ok(result) => { Ok(result) },
			Err(_) if liquidity_assets_amount == 0_u128 => {
				Ok((current_time, U256::from(0), U256::from(0)))
			}
			Err(_) => {
				let pool_promotion_start = PoolPromotionStart::<T>::get(liquidity_asset_id);

				let user_work_total = Self::calculate_work(
					liquidity_assets_amount,
					current_time,
					pool_promotion_start,
					U256::from(0),
					liquidity_assets_amount.into(),
				)?;
				let time_passed = current_time
					.checked_sub(pool_promotion_start)
					.ok_or_else(|| DispatchError::from(Error::<T>::PastTimeCalculation))?;
				let user_missing_at_checkpoint = Self::calculate_missing_at_checkpoint(
					time_passed,
					Balance::try_from(0).unwrap(),
					liquidity_assets_amount.into(),
				)?;

				LiquidityMiningUser::<T>::try_get((&user, &liquidity_asset_id))
					.unwrap_or_else(|_| (0, U256::from(0), U256::from(0)));

				Ok((current_time, user_work_total, user_missing_at_checkpoint))
			}
		}


```

Comment (mateuszaaa) [reply pallets/xyk/src/lib.rs:656]: above is a little bit more rust way

**Review by mateuszaaa — CHANGES_REQUESTED** [review]
first comments

File: pallets/xyk/src/lib.rs [hunk 4]
```diff
@@ -569,39 +592,100 @@ pub mod pallet {
...
+				.checked_sub(pool_promotion_start)
+				.ok_or_else(|| DispatchError::from(Error::<T>::PastTimeCalculation))?;
+			let user_missing_at_checkpoint = Self::calculate_missing_at_checkpoint(
+				time_passed,
+				Balance::try_from(0).unwrap(),
+				liquidity_assets_amount.into(),
+			)?;
 
-		Ok(available_rewards_for_pool)
+			LiquidityMiningUser::<T>::try_get((&user, &liquidity_asset_id))
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:652]: use regular get instead, it will retrun Default(zero) value  automatically

File: pallets/xyk/src/lib.rs [hunk 5]
```diff
@@ -569,39 +592,100 @@ pub mod pallet {
...
+			PoolPromotionStart::<T>::contains_key(liquidity_asset_id),
+			Error::<T>::NotAPromotedPool
+		);
+		let current_time: u32 =
+			<frame_system::Pallet<T>>::block_number().saturated_into::<u32>() / TIMEBLOCKRATIO;
+		let work_user = Self::calculate_work_user(user.clone(), liquidity_asset_id, current_time)?;
+		let work_pool = Self::calculate_work_pool(liquidity_asset_id, current_time)?;
+
+		let burned_not_claimed_rewards =
+			LiquidityMiningUserToBeClaimed::<T>::try_get((user, &liquidity_asset_id))
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:673]: use regular ::get instead, zeros will be returned as default

File: pallets/xyk/src/lib.rs [hunk 6]
```diff
@@ -756,30 +818,23 @@ impl<T: Config> Pallet<T> {
...
-		LiquidityMiningUser::<T>::insert(
-			(&user, &liquidity_asset_id),
-			(current_time, user_work_total, user_missing_at_checkpoint),
-		);
-
 		let (
 			pool_last_checkpoint,
 			_pool_cummulative_work_in_last_checkpoint,
 			pool_missing_at_last_checkpoint,
 		) = LiquidityMiningPool::<T>::try_get(&liquidity_asset_id)
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:825]: use ::get instead

File: pallets/xyk/src/lib.rs [hunk 7]
```diff
@@ -619,44 +703,28 @@ impl<T: Config> Pallet<T> {
...
 		let asymptote_u256: U256 = asymptote.into();
 		let cummulative_work_new_max_possible: U256 = asymptote_u256
 			.checked_mul(U256::from(time_passed))
 			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))?;
 		let base = missing_at_last_checkpoint
-			.checked_mul(U256::from(11))
-			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))?;
+			.checked_mul(U256::from(106))
+			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))? /
+			U256::from(6);
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:724]: avoid using magic numbers, instead create constants at the top of the file

File: pallets/xyk/src/lib.rs [hunk 8]
```diff
@@ -619,44 +703,28 @@ impl<T: Config> Pallet<T> {
...
 
 		let precision: u32 = 10000;
-		let q_pow = Self::calculate_q_pow(1.1, time_passed);
+		let q_pow = Self::calculate_q_pow(Q, time_passed);
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:727]: so q_pow will be removed in the future ? (not yet i mean)

File: pallets/xyk/src/lib.rs [hunk 9]
```diff
@@ -863,29 +919,23 @@ impl<T: Config> Pallet<T> {
 			),
 		);
 
-		let mut rewards_claimed: i128 = 0;
-		if LiquidityMiningUserClaimed::<T>::contains_key((user.clone(), &liquidity_asset_id)) {
-			rewards_claimed =
-				Self::liquidity_mining_user_claimed((user.clone(), &liquidity_asset_id));
+		let mut rewards_to_be_claimed: u128 = 0;
+		if LiquidityMiningUserToBeClaimed::<T>::contains_key((user.clone(), &liquidity_asset_id)) {
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:923]: use ::get instead contains key, it will either return value or 0 by default

File: pallets/xyk/src/lib.rs [hunk 10]
```diff
@@ -1940,11 +1987,14 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
 			second_asset_amount.into(),
 			ExistenceRequirement::KeepAlive,
 		)?;
-		Pallet::<T>::set_liquidity_minting_checkpoint(
-			sender.clone(),
-			liquidity_asset_id,
-			liquidity_assets_minted,
-		)?;
+		// Liquidity minting functions not triggered on not promoted pool
+		if <T as Config>::PoolPromoteApi::get_pool_rewards(liquidity_asset_id).is_some() {
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:1991]: looks like we never read value returned by get_pool_rewards, the only thing we are interested is if it is some or none. Then api should be modified to only expose that boolean information

File: pallets/xyk/src/lib.rs [hunk 11]
```diff
@@ -2167,41 +2225,87 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+			let work_to_burn = multiply_by_rational(
+				Balance::try_from(user_work_total)?,
+				rewards_to_burn,
+				current_rewards,
+			)
+			.map_err(|_| Error::<T>::UnexpectedFailure)?;
+
+			LiquidityMiningUser::<T>::insert(
+				(&user, &liquidity_asset_id),
+				(current_time, user_work_total - work_to_burn, user_missing_at_checkpoint),
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2271]: please use checked operator, so in future noone will have to think if its gonna explode ;)

File: pallets/xyk/src/lib.rs [hunk 12]
```diff
@@ -2167,41 +2225,87 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+			);
+			LiquidityMiningPool::<T>::insert(
+				&liquidity_asset_id,
+				(current_time, pool_work_total - work_to_burn, pool_missing_at_checkpoint),
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2275]: please use checked operator, so in future noone will have to think if its gonna explode ;)

**Review by mateuszaaa — CHANGES_REQUESTED** [review]
I went thorough rust side of your PR and document itself i haven't checked the tests values manually. Generally i would rather create dedicated pallet for rewards handling as it is separate functionality from the DEX itself with dedicated API to interact between those(similarly as you crated API for issuance <-> xyk interactions). It shouldnt be very hard to separate it...

File: pallets/xyk/src/lib.rs (hunk 10 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:1991]: we use the value at xyk 687:
let available_rewards_for_pool: U256 = U256::from(
			<T as Config>::PoolPromoteApi::get_pool_rewards(liquidity_asset_id)
				.ok_or_else(|| DispatchError::from(Error::<T>::NotAPromotedPool))?,
		);

File: pallets/xyk/src/lib.rs (hunk 3 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:663]: modified to a little bit more rust way

File: pallets/xyk/src/lib.rs (hunk 7 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:731]: 106 is hard number in formula, not a constant. Added comment to explain formula

File: pallets/xyk/src/lib.rs (hunk 8 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:734]: not in close future, passed to auditor to deep check used pow fn

File: pallets/xyk/src/lib.rs (hunk 11 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:2283]: let new_work_user = user_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			let new_work_pool = pool_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			
			LiquidityMiningUser::<T>::insert(
				(&user, &liquidity_asset_id),
				(current_time, new_work_user, user_missing_at_checkpoint),
			);
			LiquidityMiningPool::<T>::insert(
				&liquidity_asset_id,
				(current_time, new_work_pool, pool_missing_at_checkpoint),
			);

File: pallets/xyk/src/lib.rs (hunk 12 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:2287]: let new_work_user = user_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			let new_work_pool = pool_work_total.checked_sub(work_to_burn).ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
			
			LiquidityMiningUser::<T>::insert(
				(&user, &liquidity_asset_id),
				(current_time, new_work_user, user_missing_at_checkpoint),
			);
			LiquidityMiningPool::<T>::insert(
				&liquidity_asset_id,
				(current_time, new_work_pool, pool_missing_at_checkpoint),
			);

File: pallets/issuance/src/lib.rs [hunk 13]
```diff
@@ -207,6 +252,20 @@ impl<T: Config> Pallet<T> {
 			issuance_config.liquidity_mining_split * current_round_issuance;
 		let staking_issuance = issuance_config.staking_split * current_round_issuance;
 
+		let promoted_pools_count = <Self as PoolPromoteApi>::len();
+
+		// TODO: what about roundings? transfer mod to next session?
```
[the remark this answers is not in the document]
Comment (stanieltron) [reply pallets/issuance/src/lib.rs:259]: decided not to be important and worth code complication

File: pallets/xyk/src/lib.rs (hunk 1 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:621]: removed

File: pallets/xyk/src/lib.rs (hunk 6 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:830]: replaced by get

File: pallets/xyk/src/lib.rs (hunk 5 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:676]: replaced by get

File: pallets/xyk/src/lib.rs (hunk 6 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs:830]: replaced by get

File: pallets/xyk/src/lib.rs (hunk 9 above)
Comment (stanieltron) [reply pallets/xyk/src/lib.rs]: replaced by get

**Review by mateuszaaa — APPROVED** [review]
lgtm

**Pull request merged**